opt: add rule to normalize @> to false when const array contains NULLs

mneverov opened the pull request:
This PR adds a normalization rule that normalize a contains expression
(@>) to false when a right argument is a constant array and it contains
NULL elements.

Fixes: #56799

Release note: None

mneverov commented:
@rytaft could you please take a look?

rytaft reviewed (commented):
Nice work! Just a few nits below.

> Reviewed 5 of 5 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57026#-:-MMqRe0L4hQmPN7VUJtA:btvspkm)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mneverov)

___
*[pkg/sql/opt/norm/general_funcs.go, line 115 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57026#-MMqQ0Og9gOUJ0fZfN1I:-MMqQ0Og9gOUJ0fZfN1J:bimzeus) ([raw file](https://github.com/cockroachdb/cockroach/blob/10a9dce2b07f6c86cc778f016c60983f0c994729/pkg/sql/opt/norm/general_funcs.go#L115)):*
> ```Go
> }
> 
> // IsConstArrayWithNulls returns true if the given constant is an array that
> ```

[nit] ... if the given scalar is a constant array that ....
___
*[pkg/sql/opt/norm/rules/contains.opt, line 7 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57026#-MMqQDwaD6xtwrM72-Fa:-MMqQDwaD6xtwrM72-Fb:b-6whw7k) ([raw file](https://github.com/cockroachdb/cockroach/blob/10a9dce2b07f6c86cc778f016c60983f0c994729/pkg/sql/opt/norm/rules/contains.opt#L7)):*
> ```
> # SimplifyContainsArrayWithNulls simplifies the Contains operator to false if
> # its right argument is a constant array that contains null elements.
> [SimplifyContainsArrayWithNulls, Normalize]
> ```

I think you can actually put this in `scalar.opt`. Doesn't seem worth creating a new file for this one rule.
___
*[pkg/sql/opt/norm/rules/contains.opt, line 8 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57026#-MMqQQsK3pYWSYzY_dVW:-MMqQQsL9cCMyb_NtnMh:b-1q1krf) ([raw file](https://github.com/cockroachdb/cockroach/blob/10a9dce2b07f6c86cc778f016c60983f0c994729/pkg/sql/opt/norm/rules/contains.opt#L8)):*
> ```
> # its right argument is a constant array that contains null elements.
> [SimplifyContainsArrayWithNulls, Normalize]
> (Contains $left:* $right:* & (IsConstArrayWithNulls $right))
> ```

you shouldn't need to bind the left argument to a variable. I think you can just use `*` alone.
___
*[pkg/sql/opt/norm/testdata/rules/contains, line 6 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57026#-MMqQcMx9kFY_7rRejtu:-MMqQcMx9kFY_7rRejtv:b-gkpvs3) ([raw file](https://github.com/cockroachdb/cockroach/blob/10a9dce2b07f6c86cc778f016c60983f0c994729/pkg/sql/opt/norm/testdata/rules/contains#L6)):*
> ```
> 
> # --------------------------------------------------
> # SimplifyContainsArrayWithNulls
> ```

You can move these tests to the `scalar` test file once you move the rule to `scalar.opt`.
___
*[pkg/sql/opt/norm/testdata/rules/contains, line 41 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57026#-MMqR04HAPea8Em1273J:-MMqR04HAPea8Em1273K:b82qk8p) ([raw file](https://github.com/cockroachdb/cockroach/blob/10a9dce2b07f6c86cc778f016c60983f0c994729/pkg/sql/opt/norm/testdata/rules/contains#L41)):*
> ```
> 
> norm expect-not=SimplifyContainsArrayWithNulls
> select b @> array[]::int[] from a
> ```

this looks like a duplicate of the previous test case
___
*[pkg/sql/opt/norm/testdata/rules/contains, line 51 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57026#-MMqR7QO-g9XrkF-bB7q:-MMqR7QP4OtPjZ_AgSCM:b-uf4ymn) ([raw file](https://github.com/cockroachdb/cockroach/blob/10a9dce2b07f6c86cc778f016c60983f0c994729/pkg/sql/opt/norm/testdata/rules/contains#L51)):*
> ```
>       └── b:1 @> ARRAY[] [as="?column?":4, outer=(1), immutable]
> 
> # insert array with null element into the table.
> ```

you don't need to insert data with these optimizer tests (this test doesn't actually do anything other than print the plan of the insert -- it doesn't execute the insert). I'd remove this.


<!-- Sent from Reviewable.io -->

mneverov reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57026)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rytaft)

___
*[pkg/sql/opt/norm/general_funcs.go, line 115 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57026#-MMqQ0Og9gOUJ0fZfN1I:-MMqihC-D3mhK44yOom-:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/10a9dce2b07f6c86cc778f016c60983f0c994729/pkg/sql/opt/norm/general_funcs.go#L115)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] ... if the given scalar is a constant array that ....
</blockquote></details>

Done
___
*[pkg/sql/opt/norm/rules/contains.opt, line 7 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57026#-MMqQDwaD6xtwrM72-Fa:-MMqj-rPB_cHwP0q4BkP:b-b9wms6) ([raw file](https://github.com/cockroachdb/cockroach/blob/10a9dce2b07f6c86cc778f016c60983f0c994729/pkg/sql/opt/norm/rules/contains.opt#L7)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I think you can actually put this in `scalar.opt`. Doesn't seem worth creating a new file for this one rule.
</blockquote></details>

I agree, but when I move the rule definition to the `scalar.opt` all three positive tests are failing: `expected to see SimplifyContainsArrayWithNulls, but was not triggered. Did see FoldArray, FoldComparison, MergeProjectWithValues`.
I updated the PR with other comment fixes, but left the rule in `contains.op` file.
Could you please help me to understand what might cause this and how can I fix this?
___
*[pkg/sql/opt/norm/rules/contains.opt, line 8 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57026#-MMqQQsK3pYWSYzY_dVW:-MMqiyHq4hJXsQm01ZSo:b-y6oakf) ([raw file](https://github.com/cockroachdb/cockroach/blob/10a9dce2b07f6c86cc778f016c60983f0c994729/pkg/sql/opt/norm/rules/contains.opt#L8)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

you shouldn't need to bind the left argument to a variable. I think you can just use `*` alone.
</blockquote></details>

Fixed, thx
___
*[pkg/sql/opt/norm/testdata/rules/contains, line 6 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57026#-MMqQcMx9kFY_7rRejtu:-MMqksJnACTunPqQ8J4p:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/10a9dce2b07f6c86cc778f016c60983f0c994729/pkg/sql/opt/norm/testdata/rules/contains#L6)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

You can move these tests to the `scalar` test file once you move the rule to `scalar.opt`.
</blockquote></details>

Done.
___
*[pkg/sql/opt/norm/testdata/rules/contains, line 41 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57026#-MMqR04HAPea8Em1273J:-MMqkiIz5b64zC3s8mZ7:b-59pjz6) ([raw file](https://github.com/cockroachdb/cockroach/blob/10a9dce2b07f6c86cc778f016c60983f0c994729/pkg/sql/opt/norm/testdata/rules/contains#L41)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

this looks like a duplicate of the previous test case
</blockquote></details>

you're right, removed.
___
*[pkg/sql/opt/norm/testdata/rules/contains, line 51 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57026#-MMqR7QO-g9XrkF-bB7q:-MMqkQ2A6bhgbLhC19Yv:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/10a9dce2b07f6c86cc778f016c60983f0c994729/pkg/sql/opt/norm/testdata/rules/contains#L51)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

you don't need to insert data with these optimizer tests (this test doesn't actually do anything other than print the plan of the insert -- it doesn't execute the insert). I'd remove this.
</blockquote></details>

Removed.


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
> Reviewed 4 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57026)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mneverov)

___
*[pkg/sql/opt/norm/rules/contains.opt, line 7 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57026#-MMqQDwaD6xtwrM72-Fa:-MMqokGB2b5KlirDOJnh:b-mzxpzf) ([raw file](https://github.com/cockroachdb/cockroach/blob/10a9dce2b07f6c86cc778f016c60983f0c994729/pkg/sql/opt/norm/rules/contains.opt#L7)):*
<details><summary><i>Previously, mneverov (Max Neverov) wrote…</i></summary><blockquote>

I agree, but when I move the rule definition to the `scalar.opt` all three positive tests are failing: `expected to see SimplifyContainsArrayWithNulls, but was not triggered. Did see FoldArray, FoldComparison, MergeProjectWithValues`.
I updated the PR with other comment fixes, but left the rule in `contains.op` file.
Could you please help me to understand what might cause this and how can I fix this?
</blockquote></details>

I think that's happening since the order that rules are applied depends on the order of rules in each file, the lexicographical ordering of files, and whether or not a rule is marked as high or low priority.

It's expected that the rules you listed above are firing, and I think it's fine that they now fire before your rule and therefore prevent if from firing.  I would guess that the one that is actually preventing your rule from firing is `FoldComparison`. If you want, you can disable it for testing purposes by updating the tests to use the directive `disable=FoldComparison` next to `norm expect=SimplifyContainsArrayWithNulls`. But you should make sure that there is at least one test where your rule fires without disabling any other rules. I would expect that to happen when there is a variable on the left hand side of the contains operator.


<!-- Sent from Reviewable.io -->

mneverov reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57026)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rytaft)

___
*[pkg/sql/opt/norm/rules/contains.opt, line 7 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57026#-MMqQDwaD6xtwrM72-Fa:-MMszf2y9U0JxwRtg6l3:b-59joth) ([raw file](https://github.com/cockroachdb/cockroach/blob/10a9dce2b07f6c86cc778f016c60983f0c994729/pkg/sql/opt/norm/rules/contains.opt#L7)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I think that's happening since the order that rules are applied depends on the order of rules in each file, the lexicographical ordering of files, and whether or not a rule is marked as high or low priority.

It's expected that the rules you listed above are firing, and I think it's fine that they now fire before your rule and therefore prevent if from firing.  I would guess that the one that is actually preventing your rule from firing is `FoldComparison`. If you want, you can disable it for testing purposes by updating the tests to use the directive `disable=FoldComparison` next to `norm expect=SimplifyContainsArrayWithNulls`. But you should make sure that there is at least one test where your rule fires without disabling any other rules. I would expect that to happen when there is a variable on the left hand side of the contains operator.
</blockquote></details>

Thanks! I'd checked the [docs](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/opt/doc.go) and there was no precedence rules mentioned. 
I added `HighPriority` to this rule as I think that doing 2 casting and one field access (in `IsConstArrayWithNulls`) is faster than doing `FoldComparison`. Please let me know if I need to modify tests instead.


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
Hi @mneverov, thinking about this more I've realized that this rule unfortunately isn't correct as written, since if the left hand side is `NULL`, the result will be `NULL`, not false:
```
demo@127.0.0.1:26257/movr> create table t (a int[]);
CREATE TABLE

Time: 3ms total (execution 3ms / network 0ms)

demo@127.0.0.1:26257/movr> insert into t values (null), ('{1, 2}');
INSERT 2

Time: 3ms total (execution 3ms / network 0ms)

demo@127.0.0.1:26257/movr> select a, a @> '{null}'::int[] as contains from t;
    a   | contains
--------+-----------
  NULL  |   NULL
  {1,2} |  false
(2 rows)

Time: 1ms total (execution 1ms / network 0ms)
```
The good thing is that NULL and false are equivalent when used as filters:
```
demo@127.0.0.1:26257/movr> select * from t where a @> '{null}'::int[];
  a
-----
(0 rows)

Time: 1ms total (execution 1ms / network 0ms)
```
There are a few options here:
1. Leave this as a scalar rule, but change it to check if the left hand side is null (e.g., using a CASE statement or other construct). This may or may not have the desired effect of simplifying the expression, depending on what other rules match after the transformation is applied.
2. Create a couple of rules in `select.opt` and `join.opt` that will perform the same transformation you've already defined if the expression is a top-level filter condition.
3. Optionally create a rule in `project.opt` that will perform the same transformation you've already defined if the expression is a projections item *and* `ExprIsNeverNull` returns true for the left side. 

Sorry this is turning out to be more complicated than I originally thought, and sorry for not realizing this earlier.

> Reviewed 2 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57026#-:-MMuGu3n4Pv28NI-01Wz:b-4e1ki8)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mneverov)

___
*[Makefile, line 1633 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57026#-MMuDCly-cLEuhMIzfub:-MMuDCly-cLEuhMIzfuc:b-fkdrzv) ([raw file](https://github.com/cockroachdb/cockroach/blob/170bb6e841f115635f336ca30809f90679fe3a50/Makefile#L1633)):*
> ```Makefile
> cleanshort:
> 	rm -rf bin/.go_protobuf_sources bin/.gw_protobuf_sources
> 	-$(GO) clean $(GOFLAGS)  $(GOMODVENDORFLAGS) -tags '$(TAGS)' -ldflags '$(LINKFLAGS)' -i github.com/cockroachdb/cockroach...
> ```

You shouldn't need to be making changes to the Makefile... perhaps you're not on the right go version? We've recently upgraded to 1.15.5.
___
*[pkg/sql/opt/norm/rules/contains.opt, line 7 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57026#-MMqQDwaD6xtwrM72-Fa:-MMuDn4XEjExiwJrYVjK:b-6f2agn) ([raw file](https://github.com/cockroachdb/cockroach/blob/10a9dce2b07f6c86cc778f016c60983f0c994729/pkg/sql/opt/norm/rules/contains.opt#L7)):*
<details><summary><i>Previously, mneverov (Max Neverov) wrote…</i></summary><blockquote>

Thanks! I'd checked the [docs](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/opt/doc.go) and there was no precedence rules mentioned. 
I added `HighPriority` to this rule as I think that doing 2 casting and one field access (in `IsConstArrayWithNulls`) is faster than doing `FoldComparison`. Please let me know if I need to modify tests instead.
</blockquote></details>

Makes sense to me - thanks! Feel free to update our docs too if you think there's a good place to include info about precedence rules.
___
*[pkg/sql/opt/norm/rules/scalar.opt, line 322 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57026#-MMuDiLpCYoMRoiO64KI:-MMuDiLpCYoMRoiO64KJ:b8mo7hg) ([raw file](https://github.com/cockroachdb/cockroach/blob/170bb6e841f115635f336ca30809f90679fe3a50/pkg/sql/opt/norm/rules/scalar.opt#L322)):*
> ```
> 
> # SimplifyContainsArrayWithNulls simplifies the Contains operator to false if
> # its right argument is a constant array that contains null elements.
> ```

[nit] add a comment about why this rule is marked `HighPriority`.
___
*[pkg/sql/opt/norm/rules/scalar.opt, line 323 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57026#-MMuEKYG0pyAhrZVPrW1:-MMuDguP9GwhaepK_i--:b-bm24vd) ([raw file](https://github.com/cockroachdb/cockroach/blob/170bb6e841f115635f336ca30809f90679fe3a50/pkg/sql/opt/norm/rules/scalar.opt#L323)):*
> ```
> # SimplifyContainsArrayWithNulls simplifies the Contains operator to false if
> # its right argument is a constant array that contains null elements.
> [SimplifyContainsArrayWithNulls, Normalize, HighPriority]
> ```

[nit] I would call this rule `FoldContainsArrayWithNulls`. I think it's more consistent with how we're naming other rules that reduce the expression to a single constant.


<!-- Sent from Reviewable.io -->

rytaft commented:
(In case this wasn't clear, I'd suggest either option (1) or options (2) and (3) -- no need to do all of them)





<!-- Sent from Reviewable.io -->

rytaft commented:
Another, possibly cleaner, option is to update the constraints code. I'll add a pointer to the code in just a second.





<!-- Sent from Reviewable.io -->

rytaft commented:
Here's the pointer: https://github.com/cockroachdb/cockroach/blob/e911d58acbb016b1e266a79440592a48b1cab71b/pkg/sql/opt/memo/constraint_builder.go#L208

You can update this case to check if `datum` is an array that `HasNulls`, and if so return a `contradiction` instead of a `notNullSpan`. 

This would be in place of option (2) above, and I think this is probably a better approach.





<!-- Sent from Reviewable.io -->

mneverov reviewed (commented):
no worries, i can learn more). I'll update the PR

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57026#-:-MMusXCKB-W8qmp7YLGo:b1jwnh6)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mneverov)



<!-- Sent from Reviewable.io -->

mgartner commented:
___
*[pkg/sql/opt/norm/rules/contains.opt, line 7 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57026#-MMqQDwaD6xtwrM72-Fa:-MMvyAl4-83mYukosDGA:bbkqfl7) ([raw file](https://github.com/cockroachdb/cockroach/blob/10a9dce2b07f6c86cc778f016c60983f0c994729/pkg/sql/opt/norm/rules/contains.opt#L7)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Makes sense to me - thanks! Feel free to update our docs too if you think there's a good place to include info about precedence rules.
</blockquote></details>

Not sure, but https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/opt/optgen/lang/doc.go might be right place to document `HighPriority`. It's not mentioned there either.


<!-- Sent from Reviewable.io -->

mneverov reviewed (commented):
hi @rytaft sorry for the delay. 
I updated the constraint_builder as you mentioned (with and without added scalar rule) and it didn't help: the query still returns `false` instead of `null`. 
I also came up with the projection rule:
```
[FoldContainsArrayWithNulls, Normalize]
(Project
    $input:*
    $projections:[
        ...
        $item:(ProjectionsItem (Contains $left:* $right:*)) &
                  (ExprIsNeverNull $left (NotNullCols $input)) &
                  (IsConstArrayWithNulls $right)
        ...
    ]
    $passthrough:*
)
=>
(False)
```
which is wrong, because i need to return a RelExpr, and not ScalarExpr. The only solution I can think of is to add another general func which will return a RelExpr instead of boolean.
So the questions are:
* with modified constraint_builder do I need to keep the scalar rule?
* is it possible to construct a RelExpr with False from the norm. rules files? 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57026#-:-MN4DINWComTHbUaXn6_:b-ytxhf)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mneverov and @rytaft)



<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
Hi @mneverov,

The projection rule you wrote above looks like it's on the right track, but as you mentioned, the replace pattern is not correct. We don't want to replace the entire `Project` expression with `False` -- we only want to replace the single projections item. The rule would look something like this (you would need to write the custom function `ReplaceProjectionsItem`):

```
[FoldContainsArrayWithNulls, Normalize]
(Project
    $input:*
    $projections:[
        ...
        $item:(ProjectionsItem (Contains $left:* $right:*)) &
                  (ExprIsNeverNull $left (NotNullCols $input)) &
                  (IsConstArrayWithNulls $right)
        ...
    ]
    $passthrough:*
)
=>
(Project
    $input
    (ReplaceProjectionsItem $projections $item (False))
    $passthrough
)
```

But thinking about this more, I'm not convinced that this rule is going to be that useful. We always need to think about whether a rule is worth the overhead of checking the match pattern, and I don't see it being worth it in this case. If we find a compelling use case then we can revisit this later.

And regarding your question about whether or not to keep the scalar rule: I think you should probably get rid of it. I can't think of a way to make it both correct and efficient.

So to summarize, I think you should just update `constraint_builder.go` and not add any new rules.

Sorry for all the back-and-forth, and thanks again for the contribution!

> Reviewed 1 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57026#-:-MMvOVMHEI7ilWIwnxwg:b-49108n)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mneverov and @rytaft)



<!-- Sent from Reviewable.io -->

mneverov committed:
opt: constraint for CONTAINS

Added a constraint for @> (contains) operator. It will be evaluated to false if
the right argument is an array that contain NULLs.
Related to #56799

Release note: None

mneverov reviewed (commented):
hi @rytaft,
thanks for your help, it was interesting to poke at optimizer)
I didn't fully get the concept of `tight` boolean, could you please check?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57026#-:-MNPxBs140MG4B-tn8h2:b9iwqrl)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mneverov and @rytaft)

___
*[Makefile, line 1633 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57026#-MMuDCly-cLEuhMIzfub:-MNPvLnJ9sCdRdShgNr5:b-cqh90r) ([raw file](https://github.com/cockroachdb/cockroach/blob/170bb6e841f115635f336ca30809f90679fe3a50/Makefile#L1633)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

You shouldn't need to be making changes to the Makefile... perhaps you're not on the right go version? We've recently upgraded to 1.15.5.
</blockquote></details>

Reverted the change, thx.
___
*[pkg/sql/opt/exec/execbuilder/testdata/inverted_index, line 785 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57026#-MNPvQLZCMZaRyQ-IwQ7:-MNPvQLZCMZaRyQ-IwQ8:bebbjtu) ([raw file](https://github.com/cockroachdb/cockroach/blob/338abcf2617f7cee298cb6001335b6c46dde5b59/pkg/sql/opt/exec/execbuilder/testdata/inverted_index#L785)):*
> ```
> 
> # Test that searching for a NULL element using the inverted index.
> # TODO(rytaft): This filter should be normalized to false (see #56799).
> ```

Removed the comment as `ARRAY[NULL]` will be optimized by `FoldArray` and `buildSingleColumnConstraintConst` wont be invoked.
___
*[pkg/sql/opt/norm/rules/scalar.opt, line 322 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57026#-MMuDiLpCYoMRoiO64KI:-MNPvqBsCZlYf0EFG1lm:byjy428) ([raw file](https://github.com/cockroachdb/cockroach/blob/170bb6e841f115635f336ca30809f90679fe3a50/pkg/sql/opt/norm/rules/scalar.opt#L322)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] add a comment about why this rule is marked `HighPriority`.
</blockquote></details>

removed the rule
___
*[pkg/sql/opt/norm/rules/scalar.opt, line 323 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57026#-MMuEKYG0pyAhrZVPrW1:-MNPvsGS6lLv_fjdPMF8:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/170bb6e841f115635f336ca30809f90679fe3a50/pkg/sql/opt/norm/rules/scalar.opt#L323)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] I would call this rule `FoldContainsArrayWithNulls`. I think it's more consistent with how we're naming other rules that reduce the expression to a single constant.
</blockquote></details>

ditto


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
Nice! This is looking good. You used `tight` correctly -- it just means that the returned constraint exactly matches the values allowed by the filter, and doesn't include any additional values.

There are just a couple of small things to fix, identified in the comments below.

> Reviewed 13 of 13 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57026#-:-MNTI3LK5R2k3Jq8wEqX:b-6g02zn)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mneverov)

___
*[pkg/sql/opt/exec/execbuilder/testdata/inverted_index, line 785 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57026#-MNPvQLZCMZaRyQ-IwQ7:-MNTFhWL92omyeu1sTs1:bffdo9j) ([raw file](https://github.com/cockroachdb/cockroach/blob/338abcf2617f7cee298cb6001335b6c46dde5b59/pkg/sql/opt/exec/execbuilder/testdata/inverted_index#L785)):*
<details><summary><i>Previously, mneverov (Max Neverov) wrote…</i></summary><blockquote>

Removed the comment as `ARRAY[NULL]` will be optimized by `FoldArray` and `buildSingleColumnConstraintConst` wont be invoked.
</blockquote></details>

Hmm it looks like it's getting invoked since the estimated row count is 0. I wonder why we're not replacing this plan with an empty values node.

It would be good to add another test like this in `opt/memo/testdata/logprops/constraints` so we can see what's happening with the logical properties.
___
*[pkg/sql/opt/memo/constraint_builder.go, line 207 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57026#-MNTEjy1Dpnqfb2LMN8F:-MNTEjy1Dpnqfb2LMN8G:b-le4bo3) ([raw file](https://github.com/cockroachdb/cockroach/blob/338abcf2617f7cee298cb6001335b6c46dde5b59/pkg/sql/opt/memo/constraint_builder.go#L207)):*
> ```Go
> 			}
> 		}
> ```

[nit] add this blank line back to be consistent with the rest of the function
___
*[pkg/sql/opt/memo/constraint_builder.go, line 216 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57026#-MNTCGQX-6kWK8ibzH0u:-MNTCGQX-6kWK8ibzH0v:bjowok7) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed78b3b70045e881d333a83809b850b261c93742/pkg/sql/opt/memo/constraint_builder.go#L216)):*
> ```Go
> 			// NULL cannot contain anything, so a non-tight, not-null span is built.
> 			return cb.notNullSpan(col), false
> 		}
> ```

Now this won't return a not-null span for arrays that do not contain nulls. I'd either remove the default case and move that logic below the switch or just use the `if arr, ok := datum.(*tree.DArray); ok {` pattern. 
___
*[pkg/sql/opt/memo/testdata/logprops/constraints, line 1442 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57026#-MNTDCjP0SmU3-nbMx1U:-MNTDCjP0SmU3-nbMx1V:bvfne71) ([raw file](https://github.com/cockroachdb/cockroach/blob/338abcf2617f7cee298cb6001335b6c46dde5b59/pkg/sql/opt/memo/testdata/logprops/constraints#L1442)):*
> ```
>  │    │    └── prune: (3)
>  │    └── filters
>  │         └── contains [type=bool, outer=(3), immutable, constraints=(/3: (/NULL - ])]
> ```

here's a case where the not-null constraint was incorrectly removed
___
*[pkg/sql/opt/optgen/lang/doc.go, line 106 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57026#-MNTDa0y7MnAWIVRTX8-:-MNTDa0y7MnAWIVRTX80:b-vzim3h) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed78b3b70045e881d333a83809b850b261c93742/pkg/sql/opt/optgen/lang/doc.go#L106)):*
> ```Go
> constructed according to the replace pattern. The replacement node is itself
> tested against every rule, and so on, until no further rules match. Rules are
> applied depends on the order of the rules in each file, the lexicographical
> ```

[nit] Rules are applied depends on the order -> The order that rules are applied depends on the order
___
*[pkg/sql/opt/optgen/lang/doc.go, line 110 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57026#-MNTE0_27MqpVuXPs5-R:-MNTE0_27MqpVuXPs5-S:b-j47r56) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed78b3b70045e881d333a83809b850b261c93742/pkg/sql/opt/optgen/lang/doc.go#L110)):*
> ```Go
> as it is depicted below:
> 
> [FoldContainsArrayWithNulls, Normalize, HighPriority]
> ```

[nit] maybe pick a name that matches an existing rule (since you removed this one :) )


<!-- Sent from Reviewable.io -->

mneverov reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57026)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rytaft)

___
*[pkg/sql/opt/exec/execbuilder/testdata/inverted_index, line 785 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57026#-MNPvQLZCMZaRyQ-IwQ7:-MNUGtC16J_KBo9CW0xO:b-qhbh2n) ([raw file](https://github.com/cockroachdb/cockroach/blob/338abcf2617f7cee298cb6001335b6c46dde5b59/pkg/sql/opt/exec/execbuilder/testdata/inverted_index#L785)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Hmm it looks like it's getting invoked since the estimated row count is 0. I wonder why we're not replacing this plan with an empty values node.

It would be good to add another test like this in `opt/memo/testdata/logprops/constraints` so we can see what's happening with the logical properties.
</blockquote></details>

Added the test, it seems i was wrong.
___
*[pkg/sql/opt/memo/constraint_builder.go, line 207 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57026#-MNTEjy1Dpnqfb2LMN8F:-MNTYlE218_hqnRYzvgw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/338abcf2617f7cee298cb6001335b6c46dde5b59/pkg/sql/opt/memo/constraint_builder.go#L207)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] add this blank line back to be consistent with the rest of the function
</blockquote></details>

Done.
___
*[pkg/sql/opt/memo/constraint_builder.go, line 216 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57026#-MNTCGQX-6kWK8ibzH0u:-MNTYL5-30Nsr_0buqop:b-tgkqkq) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed78b3b70045e881d333a83809b850b261c93742/pkg/sql/opt/memo/constraint_builder.go#L216)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Now this won't return a not-null span for arrays that do not contain nulls. I'd either remove the default case and move that logic below the switch or just use the `if arr, ok := datum.(*tree.DArray); ok {` pattern. 
</blockquote></details>

:man_facepalming: you're right, fixed. This also fixed all the test files
___
*[pkg/sql/opt/memo/testdata/logprops/constraints, line 1442 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57026#-MNTDCjP0SmU3-nbMx1U:-MNUD9WY7_F-BLj1UrDR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/338abcf2617f7cee298cb6001335b6c46dde5b59/pkg/sql/opt/memo/testdata/logprops/constraints#L1442)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

here's a case where the not-null constraint was incorrectly removed
</blockquote></details>

Done.
___
*[pkg/sql/opt/optgen/lang/doc.go, line 106 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57026#-MNTDa0y7MnAWIVRTX8-:-MNTXh7d7ukYlon88F3O:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed78b3b70045e881d333a83809b850b261c93742/pkg/sql/opt/optgen/lang/doc.go#L106)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] Rules are applied depends on the order -> The order that rules are applied depends on the order
</blockquote></details>

Done.
___
*[pkg/sql/opt/optgen/lang/doc.go, line 110 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57026#-MNTE0_27MqpVuXPs5-R:-MNTXgwgC7wXmUp6O298:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed78b3b70045e881d333a83809b850b261c93742/pkg/sql/opt/optgen/lang/doc.go#L110)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] maybe pick a name that matches an existing rule (since you removed this one :) )
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Thanks for the contribution!

bors r+

> Reviewed 9 of 9 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57026#-:-MNYog-v2WlJh1QAGhF2:b-f04hy0)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mneverov)

___
*[pkg/sql/opt/exec/execbuilder/testdata/inverted_index, line 785 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57026#-MNPvQLZCMZaRyQ-IwQ7:-MNYpWOtCN9cEVXICfbg:by3nyd0) ([raw file](https://github.com/cockroachdb/cockroach/blob/338abcf2617f7cee298cb6001335b6c46dde5b59/pkg/sql/opt/exec/execbuilder/testdata/inverted_index#L785)):*
<details><summary><i>Previously, mneverov (Max Neverov) wrote…</i></summary><blockquote>

Added the test, it seems i was wrong.
</blockquote></details>

Thanks! It looks like the reason we don't replace this with an empty Values node is that the contains operator is not leak proof: https://github.com/cockroachdb/cockroach/blob/8023f54723e0b10dad13301d58490db40582e482/pkg/sql/opt/norm/factory.go#L274


<!-- Sent from Reviewable.io -->
